### ISSUE TITLE
May be there are errors in WaveNet code

### ISSUE THREAD

**Issue by azraelkuan** [Rayhane-mamah/Tacotron-2#63]
https://github.com/Rayhane-mamah/Tacotron-2/blob/e2f9780b18648b8d0b671e1c85dd95d20d6f8553/wavenet_vocoder/models/wavenet.py#L194-L197

in the line 194:
`tf.reduce_max` should be `tf.argmax`

in the line 196, 197
`y_hat_log, y_log` are Tensors, so we could not use the numpy function.

also there are similar errors in:
https://github.com/Rayhane-mamah/Tacotron-2/blob/e2f9780b18648b8d0b671e1c85dd95d20d6f8553/wavenet_vocoder/models/wavenet.py#L204-L206
and in eval process.

**Comment by a3626a** [comment]
https://github.com/Rayhane-mamah/Tacotron-2/blob/master/wavenet_vocoder/util.py is tensorflow compatible utility.

Therefore, 
> in the line 196, 197
y_hat_log, y_log are Tensors, so we could not use the numpy function.

is no problem.

For example
https://github.com/Rayhane-mamah/Tacotron-2/blob/ec0cd50d24d7f5561f696a78d0cd47d5a2a14a43/wavenet_vocoder/util.py#L128-L132

By the way, I agree with this :
> in the line 194:
tf.reduce_max should be tf.argmax

**Comment by azraelkuan** [comment]
@a3626a i forget to update the util.py and other problem:
https://github.com/Rayhane-mamah/Tacotron-2/blob/e2f9780b18648b8d0b671e1c85dd95d20d6f8553/wavenet_vocoder/models/wavenet.py#L567-L569
the function `tf.multinomial` accepts unnormlized log-probabilities for all classes, so i think we should not do softmax in the incremental step.

And the sampled wavs is not good as the train step, what problems?https://github.com/Rayhane-mamah/Tacotron-2/issues/57

**Comment by begeekmyfriend** [comment]
@Rayhane-mamah Would you like to merge another parallel wavenet version which might be faster during inference? https://github.com/andabi/parallel-wavenet-vocoder

**Comment by Rayhane-mamah** [comment]
Hello @azraelkuan and @a3626a you are right, I hopefully fixed all bugs in the code and here's a quantized wavenet sample recreated using Tacotron (160k steps) and quantized Wavenet (600k steps):

[wavenet-northandsouth_01_f000005.wav.tar.gz](https://github.com/Rayhane-mamah/Tacotron-2/files/2259718/wavenet-northandsouth_01_f000005.wav.tar.gz)

@begeekmyfriend  I don't have plans on making a parallel wavenet implementation for this work as it has not been discussed in the paper. This repo is only a base to build on top of so feel free to work on making wavenet real time :) I will be happy to support you if needed! 

By the way, feel free to join [tacotron-2 slack](https://github.com/Rayhane-mamah/Tacotron-2/issues/117) for discussions about how to improve on top of this work.
